runtime/shared_common/SCImplementedAPI.cpp

```diff
@@ -770,7 +792,9 @@ j9shr_classStoreTransaction_nextSharedClassForCompare(void * tobj)
 
 	Trc_SHR_API_j9shr_nextSharedClassForCompare_Entry(currentThread);
 
-	if (obj->transactionState != TSTATE_ENTER_WRITEMUTEX) {
+	if ((obj->transactionState != TSTATE_ENTER_WRITEMUTEX)
+		&& ((0 == obj->cacheFullFlags) || (obj->transactionState != TSTATE_ENTER_SEGMENTMUTEX))
```

Comment:
Why the check for TSTATE_ENTER_SEGMENTMUTEX?

Reply:
findNextROMClass() asserts the segmentMutex is being held. We should continue to findNextROMClass() in j9shr_classStoreTransaction_nextSharedClassForCompare either:
1) Cache is not full, write mutex is entered, or
2) Cache is full, segmentMutex is entered.

With this change, we now return from j9shr_classStoreTransaction_start() after acquiring the
segment mutex but before acquiring the cache write mutex if the cache is
full.

Reply:
What I was thinking is that it seems redundant to check both cacheFullFlags and TSTATE_ENTER_SEGMENTMUTEX. If cacheFullFlags != 0 then transactionState == TSTATE_ENTER_SEGMENTMUTEX

Reply:
Yes, the check for TSTATE_ENTER_SEGMENTMUTEX can be removed and it is removed.